File: qbopt/analysis/effects.py
"""Conservative memory effects not described by explicit MIR write ranges."""

from qbopt.model import mir

_TRAPS = frozenset({mir.Kind.FCHECK, mir.Kind.DIV, mir.Kind.REM, mir.Kind.DIVMOD})


def unmodeled_write(op: mir.Op) -> bool:
    """Missing call effects are unknown, not a proof that the call is readonly."""
    return (op.barrier and not op.memory_complete) or (op.kind is mir.Kind.CALL and not op.stores)


def exposes_memory(op: mir.Op, handles_errors: bool) -> bool:
    """Whether a trap here can reach an ON ERROR handler that reads memory.

    Without a handler a trappable error ends the program, and nothing reads
    memory after that.
    """
    return handles_errors and (op.floating is not None or op.kind in _TRAPS)

File: crates/llrm-core/src/analysis/effects.rs
//! Port of `qbopt/analysis/effects.py`: conservative memory effects not
//! described by explicit MIR write ranges.

use crate::model::mir::{Kind, Op};

const _TRAPS: [Kind; 5] = [Kind::Fcheck, Kind::Div, Kind::Rem, Kind::Divmod, Kind::FixedDiv];

/// Whether an operation may write beyond its explicit MIR store ranges.
pub fn unmodeled_write(op: &Op) -> bool {
    (op.barrier() || op.kind == Kind::Call) && !op.memory_complete
}

/// Whether an operation may read beyond its explicit MIR load ranges.
pub fn unmodeled_read(op: &Op) -> bool {
    (op.barrier() || op.kind == Kind::Call) && !op.memory_complete
}

/// Whether a trap here can reach an ON ERROR handler that reads memory.
pub fn exposes_memory(op: &Op, handles_errors: bool) -> bool {
    handles_errors && (op.floating.is_some() || _TRAPS.contains(&op.kind))
}
